features/api-layer/src/main/java/org/opennms/features/apilayer/pollers/PollerRequestBuilderImpl.java

```diff
@@ -0,0 +1,178 @@
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.opennms.integration.api.v1.pollers.PollerRequestBuilder;
+import org.opennms.integration.api.v1.pollers.PollerResult;
+import org.opennms.integration.api.v1.pollers.Status;
+import org.opennms.netmgt.poller.LocationAwarePollerClient;
+import org.opennms.netmgt.poller.MonitoredService;
+import org.opennms.netmgt.poller.PollStatus;
+import org.opennms.netmgt.poller.PollerResponse;
+import org.opennms.netmgt.poller.support.SimpleMonitoredService;
+
+/**
+ * Builder Implementation for {@link org.opennms.integration.api.v1.pollers.ServicePollerClient}
+ */
+public class PollerRequestBuilderImpl implements PollerRequestBuilder {
+
+    public PollerRequestBuilderImpl(LocationAwarePollerClient pollerClient) {
+        this.pollerClient = pollerClient;
+    }
+
+    private String className;
+
+    private InetAddress address;
+
+    private String serviceName;
+
+    private Long ttlInMs;
+
+    private Map<String, String> attributes = new HashMap<>();
+
+    private final LocationAwarePollerClient pollerClient;
+
+
+    @Override
+    public PollerRequestBuilder withPollerClassName(String className) {
+        this.className = className;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAddress(InetAddress address) {
+        this.address = address;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withServiceName(String serviceName) {
+        this.serviceName = serviceName;
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAttribute(String key, String value) {
+        this.attributes.put(key, value);
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withAttributes(Map<String, String> attributes) {
+        this.attributes.putAll(attributes);
+        return this;
+    }
+
+    @Override
+    public PollerRequestBuilder withTimeToLive(Long ttlInMs) {
+        this.ttlInMs = ttlInMs;
+        return this;
+    }
+
+    @Override
+    public CompletableFuture<PollerResult> execute() {
+        Map<String, Object> props = new HashMap<>();
+        attributes.forEach(props::put);
+        MonitoredService service = new SimpleMonitoredService(address, serviceName);
+        CompletableFuture<PollerResponse> future = pollerClient.poll()
+                .withService(service)
+                .withMonitorClassName(className)
+                .withTimeToLive(ttlInMs)
+                .withAttributes(props)
+                .execute();
+        // convert the response to PollResult.
+        CompletableFuture<PollerResult> result = new CompletableFuture<>();
+        PollStatus pollStatus;
+        try {
+            PollerResponse pollerResponse = future.get();
+            pollStatus = pollerResponse.getPollStatus();
+            switch (Status.valueOf(pollStatus.getStatusName())) {
```

Comment:
Move this logic in the switch statement into the PollerResultImpl class.

Then you can just do:
```result.complete(new PollerResultImpl(pollStatus));```

Could also use this simplified constructor:
```
private PollerResultImpl(PollStatus pollStatus) {
            status = Enums.getIfPresent(Status.class, pollStatus.getStatusName()).or(Status.Unknown);
            if(status == Status.Up) {
                properties = pollStatus.getProperties();
            } else {
                reason = pollStatus.getReason();
            }
        }
```

Reply:
Thanks for this. really made easier.